<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#EDA" data-toc-modified-id="EDA-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>EDA</a></span><ul class="toc-item"><li><span><a href="#Covers" data-toc-modified-id="Covers-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Covers</a></span></li><li><span><a href="#Lyrics" data-toc-modified-id="Lyrics-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>Lyrics</a></span></li><li><span><a href="#Meta" data-toc-modified-id="Meta-1.3"><span class="toc-item-num">1.3&nbsp;&nbsp;</span>Meta</a></span></li><li><span><a href="#Объединённый-датафрейм" data-toc-modified-id="Объединённый-датафрейм-1.4"><span class="toc-item-num">1.4&nbsp;&nbsp;</span>Объединённый датафрейм</a></span></li></ul></li><li><span><a href="#Предобработка-данных" data-toc-modified-id="Предобработка-данных-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Предобработка данных</a></span></li><li><span><a href="#Обучение-моделей" data-toc-modified-id="Обучение-моделей-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Обучение моделей</a></span><ul class="toc-item"><li><span><a href="#Baseline" data-toc-modified-id="Baseline-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Baseline</a></span></li><li><span><a href="#Дерево" data-toc-modified-id="Дерево-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Дерево</a></span></li></ul></li><li><span><a href="#Случайный-лес" data-toc-modified-id="Случайный-лес-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Случайный лес</a></span><ul class="toc-item"><li><span><a href="#Регрессия" data-toc-modified-id="Регрессия-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>Регрессия</a></span></li><li><span><a href="#CatBoost" data-toc-modified-id="CatBoost-4.2"><span class="toc-item-num">4.2&nbsp;&nbsp;</span>CatBoost</a></span></li></ul></li><li><span><a href="#AdaBoost" data-toc-modified-id="AdaBoost-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>AdaBoost</a></span><ul class="toc-item"><li><span><a href="#Stacking" data-toc-modified-id="Stacking-5.1"><span class="toc-item-num">5.1&nbsp;&nbsp;</span>Stacking</a></span></li><li><span><a href="#Тестирование" data-toc-modified-id="Тестирование-5.2"><span class="toc-item-num">5.2&nbsp;&nbsp;</span>Тестирование</a></span></li></ul></li></ul></div>

# Хакатон. Обнаружение каверов музыкальных треков

### Входные данные:
Обнаружение треков каверов - важная продуктовая задача, которая может значительно улучшить качество рекомендаций музыкального сервиса и повысить счастье наших пользователей. Если мы умеем с высокой точностью классифицировать каверы и связывать их между собой, то можно предложить пользователю новые возможности для управления потоком треков. Например:

- по желанию пользователя можем полностью исключить каверы из рекомендаций;
- показать все каверы на любимый трек пользователя;
- контролировать долю каверов в ленте пользователя.


### Описание данных

#### Разметка каверов

Файл covers.json содержит разметку каверов, сделанную редакторами сервиса:

- track_id - уникальный идентификатор трека;
- track_remake_type - метка, присвоенная редакторами. Может принимать значения ORIGINAL и COVER;
- original_track_id - уникальный идентификатор исходного трека.

<aside>
💡 Обратите внимание, что не для всех каверов известны идентификаторы исходных треков!!!

</aside>

#### Метаинформация

- track_id - уникальный идентификатор трека;
- dttm - первая дата появления информации о треке;
- title - название трека;
- language - язык исполнения;
- isrc - международный уникальный идентификатор трека;
- genres - жанры;
- duration - длительность трека;

#### Текст песен

- track_id - уникальный идентификатор трека;
- lyricId - уникальный идентификатор текста;
- text - текст трека.

### Цель проекта

В этом хакатоне вам предлагается разработать решение, которое:

- может классифицировать треки по признаку кавер-некавер;
- связывать (группировать) каверы и исходный трек;
- находит исходный трек в цепочке каверов.

Совсем не обязательно, чтобы ваше решение умело делать все три функции. Будет уже здорово, если вы сможете придумать решение хотя бы для одной задачи.

Для решения этой задачи вы можете использовать любые открытые источники данных и собирать необходимую вам информацию, соблюдая правила использования сервисов, которые эту информацию предоставляют.

## План проекта

Предстоит построить прототип модели машинного обучения. Его разработка состоит из нескольких этапов:

1. Исследовательский анализ данных.
- провести ознакомление с данными;
- по необходимости изменить тип данных;
- определить категориальные признаки;
- проверить данные на пропуски;
- заполнить пропуски по необходимости;
- выявить влияние категориальных признаков на целевой;
- выявить корреляцию между признаками;
- объединить данные и посмотреть на зависимости;
- сделать выводы по результатам анализа данных.

2. Подготовка данных для моделей.
- объединить все таблицы в одну;
- при необходимости добавить признаки;
- провести преобразование категориальных признаков в численные техникой ОНЕ;
- разделить данные на 2 выборки обучающую и тестовую в пропорции 3:1
- провести стандартизацию данных;
- избавиться от дисбалланса признаков
- провести оценку важности признаков
- провести анализ матрицы ошибок

3. Обучение моделей.
- Baseline;
- LogisticRegression;
- RandomForestClassifier;
- CatBoostClassifier
- применить стэкинг

4. Отчет по проведенной работе.

In [1]:
!pip install langdetect

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 981.5/981.5 kB 11.9 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
  Created wheel for langdetect: filename=langdetect-1.0.9-py3-none-any.whl size=993224 sha256=e1d2970294aca02e5e4018b34590cfb2b6ef70e04db6980671443c1cdfcdc511
  Stored in directory: /root/.cache/pip/wheels/95/03/7d/59ea870c70ce4e5a370638b5462a7711ab78fba2f655d05106
Successfully built langdetect


In [2]:
# from google.colab import drive
# drive.mount('/content/drive')

In [3]:
!pip install phik

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 679.5/679.5 kB 9.6 MB/s eta 0:00:00


In [4]:
!pip install catboost

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 98.7/98.7 MB 9.4 MB/s eta 0:00:00


In [5]:
!pip install optuna

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 409.6/409.6 kB 6.8 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 226.8/226.8 kB 24.3 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 78.7/78.7 kB 8.1 MB/s eta 0:00:00


In [6]:
!pip install gensim

In [7]:
import os
import pandas as pd
import json
import matplotlib.pyplot as plt
import seaborn as sns
import numpy as np
import phik
import optuna

In [8]:
from langdetect import detect
from sklearn.model_selection import train_test_split, GridSearchCV, cross_val_score
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier, AdaBoostClassifier, StackingClassifier
from sklearn.linear_model import LogisticRegression
from lightgbm import LGBMClassifier
from catboost import CatBoostClassifier
from sklearn.preprocessing import LabelEncoder, OneHotEncoder, StandardScaler, OrdinalEncoder
from sklearn.dummy import DummyClassifier
from sklearn.metrics import f1_score, make_scorer
from sklearn.feature_extraction.text import TfidfVectorizer
from scipy.sparse import csr_matrix
from sklearn.decomposition import PCA
from tqdm import tqdm
from concurrent.futures import ProcessPoolExecutor
tqdm.pandas()

In [9]:
RANDOM_STATE = 42

Прочитаем предоставленные данные

In [10]:
def json_opening(path):
  data = []

  with open(path, encoding='utf-8') as datafile:
    for line in datafile:
        data.append(json.loads(line))

  return pd.DataFrame(data)


In [11]:
df_covers = json_opening('/covers.json')
df_lyrics = json_opening('/lyrics.json')
df_meta = json_opening('/meta.json')

## EDA

### Covers

Проведем изучение датасета `Covers`

In [12]:
df_covers.head()

original_track_id                          track_id  \
0  eeb69a3cb92300456b6a5f4162093851  eeb69a3cb92300456b6a5f4162093851   
1  fe7ee8fc1959cc7214fa21c4840dff0a  fe7ee8fc1959cc7214fa21c4840dff0a   
2  cd89fef7ffdd490db800357f47722b20  cd89fef7ffdd490db800357f47722b20   
3  995665640dc319973d3173a74a03860c  995665640dc319973d3173a74a03860c   
4                              None  d6288499d0083cc34e60a077b7c4b3e1   

  track_remake_type  
0          ORIGINAL  
1          ORIGINAL  
2          ORIGINAL  
3          ORIGINAL  
4             COVER

In [13]:
df_covers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 71597 entries, 0 to 71596
Data columns (total 3 columns):
 #   Column             Non-Null Count  Dtype 
---  ------             --------------  ----- 
 0   original_track_id  4821 non-null   object
 1   track_id           71597 non-null  object
 2   track_remake_type  71597 non-null  object
dtypes: object(3)
memory usage: 1.6+ MB


In [14]:
df_covers.describe()

original_track_id                          track_id  \
count                               4821                             71597   
unique                              4468                             71597   
top     f3e5e9db509f9b6f6d93b9c38e4e1324  eeb69a3cb92300456b6a5f4162093851   
freq                                  38                                 1   

       track_remake_type  
count              71597  
unique                 2  
top                COVER  
freq               67304

In [15]:
df_covers.isna().sum()

original_track_id    66776
track_id                 0
track_remake_type        0
dtype: int64

Все фичи имеют корректные названия. Огромное количество нанов в столбце `original_track_id`.

Всего треков в датасете сильно больше, чем оригинальных треков что закономерно.

Замечаем, что в столбце `original_track_id`	есть почти 400 дубликатов. Видимо это как раз те оригиналы, на которых есть больше чем 1 кавер. Это странно и интересно, ведь информация в этом столбце это не айди трека, которого кавер, а исключительно айди оригинальных треков. Чтобы разобраться, откуда дубликаты, сначала предположим, что `original_track_id` есть только у тех треков, которые с лейблом `ORIGINAL`. Посмотрим, есть ли такие id оригиналов, у которых есть метка `COVER`.

In [16]:
df_covers.groupby('track_remake_type')['original_track_id'].count()

track_remake_type
COVER        528
ORIGINAL    4293
Name: original_track_id, dtype: int64

Так и есть, получаем 528 треков в `original_track_id` с меткой `COVER`. Это даже больше, чем дубликатов в столбце. Оставим только эти треки и посмотрим, есть ли среди них дубликаты

In [17]:
duplicates = df_covers[df_covers['track_remake_type'] == 'COVER']['original_track_id'].duplicated(keep=False)
duplicate_tracks = df_covers[df_covers['track_remake_type'] == 'COVER'][duplicates]
duplicate_tracks['original_track_id'].count()

330

Получается 330 дубликатов. Это почти столько же, сколько Дубликатов по этому столбцу всего (353). Итак, теперь понятно, что всё-таки у нас есть несколько каверов, для которых мы уже знаем их id оригинального трека.

Ещё было бы не лишним посмотреть, действительно ли все оригинальные треки имеют свой `original_id`.

In [18]:
df_covers[df_covers['track_remake_type'] == 'ORIGINAL']['original_track_id'].isna().sum()

0

Да, так и есть. Все треки без `original_id` являются каверами.

In [19]:
df_covers.duplicated().sum()

0

При этом полные дубликаты в датасете тоже отсутствуют.

### Lyrics

In [20]:
df_lyrics.head()

lyricId  \
0  a951f9504e89759e9d23039b7b17ec14   
1  0c749bc3f01eb8e6cf986fa14ccfc585   
2  e2c8830fbc86e5964478243099eec23a   
3  e2c8830fbc86e5964478243099eec23a   
4  7624653ca8522ba93470843c74961b7d   

                                                text  \
0  Живу сейчас обломами, обломками не той любви\n...   
1  Tell me your fable\nA fable\nTell me your fabl...   
2  You're ashamed about all your fears and doubts...   
3  You're ashamed about all your fears and doubts...   
4  You showed him all the best of you,\nBut I'm a...   

                           track_id  
0  1c4b1230f937e4c548ff732523214dcd  
1  0faea89b0d7d6235b5b74def72511bd8  
2  9c6dc41d5ccd9968d07f055da5d8f741  
3  bfd04a73e9cffdf0e282c92219a86ea1  
4  8d70930d09cd239c948408d1317d8659

In [21]:
df_lyrics.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11414 entries, 0 to 11413
Data columns (total 3 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   lyricId   11414 non-null  object
 1   text      11414 non-null  object
 2   track_id  11414 non-null  object
dtypes: object(3)
memory usage: 267.6+ KB


In [22]:
df_lyrics.describe()

lyricId  \
count                              11414   
unique                             10915   
top     28fbafaf626c497d0d4d933b4f0d3f5b   
freq                                   9   

                                                     text  \
count                                               11414   
unique                                              10644   
top     Twenty-five years and my life is still\nI'm tr...   
freq                                                   15   

                                track_id  
count                              11414  
unique                             10277  
top     b891000882f4a612190142861b5c038b  
freq                                  24

In [23]:
df_lyrics.isna().sum()

lyricId     0
text        0
track_id    0
dtype: int64

Пропусков нет, все столбцы с правильным названием. Переименуем `lyricId` в `lyric_id`.

Замечаем, что присутствуют дубликаты в стобце `track_id`, который должен быть уникальным. В результате анализа обнаружили, что это связано с тем, что некоторые тексты песен различаются лишь несколькими символами, хотя по сути являются одинаковыми. Избавимся от этих дубликатов.

In [24]:
df_lyrics.drop_duplicates(subset='track_id', keep='first', inplace=True)

In [25]:
df_lyrics.rename(columns={'lyricId': 'lyric_id'}, inplace=True)

In [26]:
df_lyrics.describe()

lyric_id  \
count                              10277   
unique                              9891   
top     6fe05ce48761f833737487cd7a499cd3   
freq                                   9   

                                                     text  \
count                                               10277   
unique                                               9852   
top     A and X, El Orfanato\nDanza Kuduro\nLucenzo El...   
freq                                                    9   

                                track_id  
count                              10277  
unique                             10277  
top     1c4b1230f937e4c548ff732523214dcd  
freq                                   1

Готово

### Meta

In [27]:
df_meta.head()

track_id          dttm                 title  \
0  c3b9d6a354ca008aa4518329aaa21380  1.639688e+12        Happy New Year   
1  c57e3d13bbbf5322584a7e92e6f1f7ff  1.637762e+12            Bad Habits   
2  955f2aafe8717908c140bf122ba4172d  1.637768e+12  Por Esa Loca Vanidad   
3  fae5a077c9956045955dde02143bd8ff  1.637768e+12          Mil Lagrimas   
4  6bede082154d34fc18d9a6744bc95bf5  1.637768e+12   Sexo Humo y Alcohol   

  language          isrc             genres  duration  
0       EN  RUB422103970            [DANCE]  161120.0  
1       EN  QZN882178276      [ELECTRONICS]  362260.0  
2     None  QZNJZ2122549  [FOLK, LATINFOLK]  260000.0  
3     None  QZNJZ2166033  [FOLK, LATINFOLK]  190000.0  
4     None  QZNJZ2122551  [FOLK, LATINFOLK]  203000.0

In [28]:
df_meta.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 71769 entries, 0 to 71768
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   track_id  71768 non-null  object 
 1   dttm      71768 non-null  float64
 2   title     71768 non-null  object 
 3   language  21969 non-null  object 
 4   isrc      71455 non-null  object 
 5   genres    71768 non-null  object 
 6   duration  71768 non-null  float64
dtypes: float64(2), object(5)
memory usage: 3.8+ MB


In [29]:
df_meta.isna().sum()

track_id        1
dttm            1
title           1
language    49800
isrc          314
genres          1
duration        1
dtype: int64

In [30]:
df_meta.sort_values(by='track_id').tail()

track_id          dttm                   title  \
35829  fffc8f02feca70f7fad88fb987b76c4e  1.633135e+12  La Luz de la Cara Roja   
36591  fffefc046267a3f6263fc2c8d0f3f871  1.633989e+12          Uska Hi Banana   
71472  ffff8f1c01d78267ce067a2507fcc8f8  1.558112e+12                No Limit   
46002  ffff9a22aa441db772206280ffb8b52e  1.423084e+12          You're a Woman   
44434                              None           NaN                    None   

      language          isrc           genres  duration  
35829     None  ARAB82100052  [ROCK, ALLROCK]  200100.0  
36591     None  QZNJX2142230           [FOLK]  185880.0  
71472     None  PLS921752714               []  175950.0  
46002     None          None            [POP]  321820.0  
44434     None          None             None       NaN

Названия правильные, дата не в неправильном формате, приведём к `datetime`.

Присутсвуют пропуски в столбцах `language` (очень много) и `isrc`.

Есть одна полностью пустая строка, удалим ее.

Язык будем восстанавливать по названию трека.

Пропуски в `isrc` никак не восстановить - удалим.

In [31]:
df_meta.dropna(how='all', inplace=True)

In [32]:
df_meta['dttm'] = pd.to_datetime(df_meta['dttm'], unit='ms')

Теперь создадим функцию исключение, чтобы `langdetector` не выдал ошибку. Применяем функцию, заполняем пропуски языка по названию треков

In [33]:
%%time
def detect_language(text):
    try:
        result = detect(text)
        return result.upper()
    except:
        return None

df_meta['language'].fillna(df_meta['title'].progress_apply(lambda x: detect_language(str(x)) if pd.notna(x) and x.strip() else None), inplace=True)

100%|██████████| 71768/71768 [18:26<00:00, 64.83it/s]

CPU times: user 16min 57s, sys: 7.4 s, total: 17min 4s
Wall time: 18min 27s


In [34]:
df_meta.isna().sum()

track_id      0
dttm          0
title         0
language     69
isrc        313
genres        0
duration      0
dtype: int64

Проверяем - готово, осталось всего 69 миссов! Отличный результат. Удаляем остатки

In [35]:
df_meta.dropna(inplace=True)

Проверим, корректно ли заполнился язык.

In [36]:
df_meta.head(30)

track_id                dttm  \
0   c3b9d6a354ca008aa4518329aaa21380 2021-12-16 21:00:00   
1   c57e3d13bbbf5322584a7e92e6f1f7ff 2021-11-24 14:00:33   
2   955f2aafe8717908c140bf122ba4172d 2021-11-24 15:37:03   
3   fae5a077c9956045955dde02143bd8ff 2021-11-24 15:37:03   
4   6bede082154d34fc18d9a6744bc95bf5 2021-11-24 15:37:03   
5   551331ee1444de6658e71933eb7f57a2 2021-11-24 15:37:03   
6   2be8bb01bb1aeef0908524dc11935a88 2021-11-24 15:37:03   
7   3883d1add3ff8b60e77548a64df7c892 2021-11-24 15:37:03   
8   59aa886979699aa7e4e2a8aed43bfe03 2021-11-24 15:37:03   
9   2a38992b7769963937246288ad89b44c 2021-11-24 15:37:03   
10  530c73cb44a3b3abd7f2b299977fbb19 2021-11-24 15:37:03   
11  c11320ee81872399e23a2670749d393a 2021-11-24 15:37:03   
12  5f95372a5ae6e6d0d7606ea11dce1ece 2021-11-24 15:37:03   
13  adfa176e2fb0b952569129a22d64de4d 2021-11-24 15:37:03   
14  a7309d9c3374e12201f3226bc0fa7980 2021-11-24 15:37:21   
15  1069db8b42f2f79b6a66ae57c6332670 2021-11-24 16:47:17   
16  f973c844c174b9eafca7fa87560c2df9 2021-11-24 16:47:18   
17  c996348ac49e0c243a4eb4cb0a9cfbbb 2021-11-24 16:57:06   
18  55a4890cc667212a048ce81b4049d1ea 2021-11-24 16:59:05   
19  f8de56ba1399dfab829176023352024d 2021-11-24 17:03:24   
20  c0d52a0b4713145bcd3b10b7e5315bfa 2021-11-24 17:03:24   
21  0cd778cff703e232d0626ee47ec8c0ae 2021-11-24 17:03:24   
22  0c0315f45a49a35901912d0084cdf03f 2021-11-24 17:03:24   
23  b07061f3d66aee20c96ed970a46c227b 2021-11-24 17:03:24   
24  6baf269773c2bff0aa080991de62115d 2021-11-24 17:03:24   
25  94093810e71074bec5a6990c1c214165 2022-10-19 21:00:00   
26  2b74d8dc05d365e51cf4ff2bcc9727f0 2021-12-09 21:00:00   
27  bdca41f86284a6284c8e48c440bcced0 2021-11-24 17:42:00   
28  988b3d43fc47fdf843b6c21de0fe3bb4 2021-11-24 17:59:32   
29  b83e2289c343f6196c08c2b48a9c689f 2021-11-24 17:59:40   

                                                title language          isrc  \
0                                      Happy New Year       EN  RUB422103970   
1                                          Bad Habits       EN  QZN882178276   
2                                Por Esa Loca Vanidad       PT  QZNJZ2122549   
3                                        Mil Lagrimas       IT  QZNJZ2166033   
4                                 Sexo Humo y Alcohol       ES  QZNJZ2122551   
5                                       La Ultima Vez       IT  QZNJZ2166034   
6                     Homenaje a Pastor de los Santos       ES  QZNJZ2122553   
7                                      Los Marginales       ES  QZNJZ2122554   
8                                      Cuenta Conmigo       ES  QZNJZ2122555   
9                                      Ahora Juego Yo       ES  QZNJZ2122556   
10                                           Mariposa       ES  QZNJZ2166035   
11  Voy a Beber / Piensas en Mi /  Por el Momento ...       ES  QZNJZ2166036   
12                                   Adicto a Tu Piel       ES  QZNJZ2166037   
13                                  Mi Amor Imposible       CA  QZNJZ2166038   
14                  Canto para Pedir Posada Cumbiamix       PT  QZNJZ2167682   
15                                         The Moment       EN  QZNJZ2173087   
16                                    Forever in Love       NO  QZNJZ2173055   
17                               Boa Noite Vizinhança       PT  QZNJZ2178383   
18                                      Tacones Rojos       ES  GBKPL2174553   
19                                  Parabéns pra Você       PT  QZNJZ2137727   
20                                            Timidez       TR  QZNJZ2137728   
21                                             Trovão       PT  QZNJZ2137729   
22                                 Inigualável Paixão       PT  QZNJZ2137730   
23                                 Pra Você Acreditar       PT  QZNJZ2137731   
24                                        Pé na Areia       PT  QZNJZ2137732   
25                                           Humsafar       HI  GX3HH2123738   
26     

Визуально всё в правильном формате, и почти всегда попадает в точку. Проверяем остатки пропусков.

In [37]:
df_meta.isna().sum()

track_id    0
dttm        0
title       0
language    0
isrc        0
genres      0
duration    0
dtype: int64

Готово! теперь посмотрим дубликаты

In [38]:
df_meta.drop(columns=['genres'], axis=1, inplace=False).duplicated().sum()

0

In [39]:
df_meta['track_id'].duplicated().sum()

0

In [40]:
df_meta['isrc'].duplicated().sum()

172

Полные дубликаты отсутствуют, `track_id` все уникальные. Присутсвуют дубликаты в `isrc`, который по идее должен быть уникальным. Скорее всего это разные релизы одного и того же трека. Удалим их.

In [41]:
df_meta.drop_duplicates(subset='isrc', keep='first', inplace=True)

In [42]:
df_meta['isrc'].duplicated().sum() #проверяем

0

Готово! Начальный анализ проведён.

**Вывод:** у нас есть три датасета, содержащих разную информацию. От пропусков избавились (кроме `original_track_id`), аномальные дубликаты удалены. Теперь можно сделать из них единый датафрейм. Уникальный `track_id` назначим индексом.

### Объединённый датафрейм

In [43]:
df = df_covers.merge(df_lyrics, on='track_id', how='outer').merge(df_meta, on='track_id', how='outer')

df.set_index('track_id', inplace=True)

In [44]:
df.head()

original_track_id  \
track_id                                                             
eeb69a3cb92300456b6a5f4162093851  eeb69a3cb92300456b6a5f4162093851   
fe7ee8fc1959cc7214fa21c4840dff0a  fe7ee8fc1959cc7214fa21c4840dff0a   
cd89fef7ffdd490db800357f47722b20  cd89fef7ffdd490db800357f47722b20   
995665640dc319973d3173a74a03860c  995665640dc319973d3173a74a03860c   
d6288499d0083cc34e60a077b7c4b3e1                              None   

                                 track_remake_type  \
track_id                                             
eeb69a3cb92300456b6a5f4162093851          ORIGINAL   
fe7ee8fc1959cc7214fa21c4840dff0a          ORIGINAL   
cd89fef7ffdd490db800357f47722b20          ORIGINAL   
995665640dc319973d3173a74a03860c          ORIGINAL   
d6288499d0083cc34e60a077b7c4b3e1             COVER   

                                                          lyric_id  \
track_id                                                             
eeb69a3cb92300456b6a5f4162093851  260f21d9f48e8de874a6e844159ddf28   
fe7ee8fc1959cc7214fa21c4840dff0a  2498827bd11eca5846270487e4960080   
cd89fef7ffdd490db800357f47722b20  5237001311d4062bf2b80de30652bf58   
995665640dc319973d3173a74a03860c  e5b1b57090b728e8d98d2b4d9b781bf4   
d6288499d0083cc34e60a077b7c4b3e1                               NaN   

                                                                               text  \
track_id                                                                              
eeb69a3cb92300456b6a5f4162093851  Left a good job in the city\nWorkin' for the m...   
fe7ee8fc1959cc7214fa21c4840dff0a  Some folks are born made to wave the flag\nOoh...   
cd89fef7ffdd490db800357f47722b20  Uno por pobre y feo, hombre\nPero antoja'o, ay...   
995665640dc319973d3173a74a03860c  Yeah!... yeah!... remember the time, baby... y...   
d6288499d0083cc34e60a077b7c4b3e1                                                NaN   

                                                dttm            title  \
track_id                                                                
eeb69a3cb92300456b6a5f4162093851 2009-11-11 21:00:00       Proud Mary   
fe7ee8fc1959cc7214fa21c4840dff0a 2009-11-11 21:00:00    Fortunate Son   
cd89fef7ffdd490db800357f47722b20 2009-09-21 20:00:00  La Camisa Negra   
995665640dc319973d3173a74a03860c 2009-11-16 21:00:00    The Way I Are   
d6288499d0083cc34e60a077b7c4b3e1 2009-09-17 20:00:00        Extraball   

                                 language          isrc  \
track_id                                                  
eeb69a3cb92300456b6a5f4162093851       EN  USFI86900049   
fe7ee8fc1959cc7214fa21c4840dff0a       EN  USFI86900065   
cd89fef7ffdd490db800357f47722b20       ES  USUL10400965   
995665640dc319973d3173a74a03860c       EN  USUM70722806   
d6288499d0083cc34e60a077b7c4b3e1       EN  FR8Q10900116   

                                                  genres  duration  
track_id                                                            
eeb69a3cb92300456b6a5f4162093851         [ROCK, ALLROCK]  187220.0  
fe7ee8fc1959cc7214fa21c4840dff0a         [ROCK, ALLROCK]  137780.0  
cd89fef7ffdd490db800357f47722b20  [POP, FOLK, LATINFOLK]  216840.0  
995665640dc319973d3173a74a03860c       [FOREIGNRAP, RAP]  179660.0  
d6288499d0083cc34e60a077b7c4b3e1           [ELECTRONICS]  212620.0

In [45]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 71767 entries, eeb69a3cb92300456b6a5f4162093851 to b0e056b8cdd428972b758e8ea8a0ce8c
Data columns (total 10 columns):
 #   Column             Non-Null Count  Dtype         
---  ------             --------------  -----         
 0   original_track_id  4821 non-null   object        
 1   track_remake_type  71597 non-null  object        
 2   lyric_id           10277 non-null  object        
 3   text               10277 non-null  object        
 4   dttm               71215 non-null  datetime64[ns]
 5   title              71215 non-null  object        
 6   language           71215 non-null  object        
 7   isrc               71215 non-null  object        
 8   genres             71215 non-null  object        
 9   duration           71215 non-null  float64       
dtypes: datetime64[ns](1), float64(1), object(8)
memory usage: 6.0+ MB


In [47]:
df.isna().sum()

original_track_id    66946
track_remake_type      170
lyric_id             61490
text                 61490
dttm                   552
title                  552
language               552
isrc                   552
genres                 552
duration               552
dtype: int64

После объединения появились пропуски. Что касается `id` оригиналов и текста - ничего удивительного, а откуда в остальных колонках? Посмотрим. Выводим хвост датафрейма, отсортированного по названию (где точно не должно по идее быть пропусков)

In [48]:
df.sort_values(by='title').tail(20)

original_track_id  \
track_id                                                             
e03f31f1286208dcb0cf1e5ee84aceff  e03f31f1286208dcb0cf1e5ee84aceff   
c0308f925cc2c6d8b70aef91e7b8f338                              None   
23a6c682503e8a9829f70b5f8925c62b                              None   
8f9dd56bdeb9ada69cd291899f52d5a8  8f9dd56bdeb9ada69cd291899f52d5a8   
571a376663b9e86159b2234841700cda                              None   
36dc45aa77dc758427f3aaab066d569b                              None   
753a6567816f1f1fd458caeaeb33ea6a  753a6567816f1f1fd458caeaeb33ea6a   
eccece56e9409fbc59cfab0a5c0d1787  eccece56e9409fbc59cfab0a5c0d1787   
137f3f182530fb4ba015ef2ce69eedfc  137f3f182530fb4ba015ef2ce69eedfc   
ac71c0295c370a352f170431e1156b0f  ac71c0295c370a352f170431e1156b0f   
c63d43ce874ea17a1e98aa4ae526af73  c63d43ce874ea17a1e98aa4ae526af73   
7ee22bc05263ac26e24e5f895e1dbc72                              None   
e710cc10abdd5b16107c71d231a52df1  e710cc10abdd5b16107c71d231a52df1   
197694c04d7d6b2426cdebad6f06b57a  197694c04d7d6b2426cdebad6f06b57a   
341c141769af816e128d3906fd5b90f8  341c141769af816e128d3906fd5b90f8   
3d4b4c7adc6b80270711517280ad22e3                              None   
176a983bf850334e5c36d965b726a6e0                              None   
953ef412c3fd8e6698219e7e486af561                              None   
872140dcd8a61120041ef75f53188e76                               NaN   
f994d35681afefa37e8448c5e90a28e9                               NaN   

                                 track_remake_type  \
track_id                                             
e03f31f1286208dcb0cf1e5ee84aceff          ORIGINAL   
c0308f925cc2c6d8b70aef91e7b8f338             COVER   
23a6c682503e8a9829f70b5f8925c62b             COVER   
8f9dd56bdeb9ada69cd291899f52d5a8          ORIGINAL   
571a376663b9e86159b2234841700cda             COVER   
36dc45aa77dc758427f3aaab066d569b             COVER   
753a6567816f1f1fd458caeaeb33ea6a          ORIGINAL   
eccece56e9409fbc59cfab0a5c0d1787          ORIGINAL   
137f3f182530fb4ba015ef2ce69eedfc          ORIGINAL   
ac71c0295c370a352f170431e1156b0f          ORIGINAL   
c63d43ce874ea17a1e98aa4ae526af73          ORIGINAL   
7ee22bc05263ac26e24e5f895e1dbc72             COVER   
e710cc10abdd5b16107c71d231a52df1          ORIGINAL   
197694c04d7d6b2426cdebad6f06b57a          ORIGINAL   
341c141769af816e128d3906fd5b90f8          ORIGINAL   
3d4b4c7adc6b80270711517280ad22e3             COVER   
176a983bf850334e5c36d965b726a6e0             COVER   
953ef412c3fd8e6698219e7e486af561             COVER   
872140dcd8a61120041ef75f53188e76               NaN   
f994d35681afefa37e8448c5e90a28e9               NaN   

                                                          lyric_id  \
track_id                                                             
e03f31f1286208dcb0cf1e5ee84aceff                               NaN   
c0308f925cc2c6d8b70aef91e7b8f338                               NaN   
23a6c682503e8a9829f70b5f8925c62b                               NaN   
8f9dd56bdeb9ada69cd291899f52d5a8                               NaN   
571a376663b9e86159b2234841700cda                               NaN   
36dc45aa77dc758427f3aaab066d569b                               NaN   
753a6567816f1f1fd458caeaeb33ea6a  8a05c998880f8ae66cbd708c0537ed1b   
eccece56e9409fbc59cfab0a5c0d1787                               NaN   
137f3f182530fb4ba015ef2ce69eedfc                               NaN   
ac71c0295c370a352f170431e1156b0f                               NaN   
c63d43ce874ea17a1e98aa4ae526af73                               NaN   
7ee22bc05263ac26e24e5f895e1dbc72                               NaN   
e710cc10abdd5b16107c71d231a52df1  cd57f5fb44ec0887dc1b30cf44b0b8e9   
197694c04d7d6b2426cdebad6f06b57a  f431426dcfeec5f2dee9516ff3e85755   
341c141769af816e128d3906fd5b90f8                               NaN   
3d4b4c7adc6b80270711517280ad22e3                               NaN   
176a983bf850334e5c36d965b726a6e0                           

Видим, что в датафрейме присутствуют строки, про которые нет никакой информации, кроме `id` трека и метки оригинал\кавер. Можно было бы заполнить эту информацию по `id` оригинального трека, но он есть только у треков с меткой `ORIGINAL`, и скорее всего такой информации просто нет в датасете. К тому же, здесь присутствуют строки вовсе без `id` оригинала, только с меткой `COVER`. Придётся удалить эти строки из датасета. Теперь посмотрим как выглядят строки с пропусками по метке кавер\оригинал.

In [49]:
df.sort_values(by='track_remake_type').tail(20)

original_track_id track_remake_type  \
track_id                                                               
1e8ca836c962598551882e689265c1c5               NaN               NaN   
1296f487d135fa2808e18accebb6554f               NaN               NaN   
5a1fce1ebb19da07487f3132b261b78e               NaN               NaN   
a33e4b17ce2c355fb3e65ee042509803               NaN               NaN   
0e2b32a929c0e09569fdd0c831e2b459               NaN               NaN   
a6bea0f1418b5e768b689604e11d111d               NaN               NaN   
fb68073731d1fb23269c9778df343711               NaN               NaN   
96437a3517f1f408381363bc8d539718               NaN               NaN   
694ddd9c20051cd893222f00497b58a7               NaN               NaN   
ac8c3425ac530ce79214dd0704a08e30               NaN               NaN   
0f74b97c7afdd09412b1147e812aad76               NaN               NaN   
7684d25699831e0e09c923ec709f602c               NaN               NaN   
39282a5bda389deac70c6c33ebd05fe7               NaN               NaN   
105e505f39a1ce3a9e277e81783a8ffc               NaN               NaN   
43bcd59d680a05b9ff28993e6375cb64               NaN               NaN   
fa1b6f8b3ba1609dfaf6bda5b180ef25               NaN               NaN   
5e16b48a01483936ae404c3b30286a02               NaN               NaN   
b8d4dc8f7abdd59df56eb39787b35c8f               NaN               NaN   
b9233cb60674ca1ebb868e1cfe8d7b5c               NaN               NaN   
b0e056b8cdd428972b758e8ea8a0ce8c               NaN               NaN   

                                                          lyric_id  \
track_id                                                             
1e8ca836c962598551882e689265c1c5  bccf43f213b3abecaa0bdcdc17f5fff7   
1296f487d135fa2808e18accebb6554f  144248a3cb888ffc4f28c4b3df9f45c1   
5a1fce1ebb19da07487f3132b261b78e  6a3fdd4107a56ad1d9d622246b6ba327   
a33e4b17ce2c355fb3e65ee042509803  bc9d77c8771c3d81aa848adc339f1fb5   
0e2b32a929c0e09569fdd0c831e2b459                               NaN   
a6bea0f1418b5e768b689604e11d111d                               NaN   
fb68073731d1fb23269c9778df343711                               NaN   
96437a3517f1f408381363bc8d539718                               NaN   
694ddd9c20051cd893222f00497b58a7                               NaN   
ac8c3425ac530ce79214dd0704a08e30                               NaN   
0f74b97c7afdd09412b1147e812aad76                               NaN   
7684d25699831e0e09c923ec709f602c                               NaN   
39282a5bda389deac70c6c33ebd05fe7                               NaN   
105e505f39a1ce3a9e277e81783a8ffc                               NaN   
43bcd59d680a05b9ff28993e6375cb64                               NaN   
fa1b6f8b3ba1609dfaf6bda5b180ef25                               NaN   
5e16b48a01483936ae404c3b30286a02                               NaN   
b8d4dc8f7abdd59df56eb39787b35c8f                               NaN   
b9233cb60674ca1ebb868e1cfe8d7b5c                               NaN   
b0e056b8cdd428972b758e8ea8a0ce8c                               NaN   

                                                                               text  \
track_id                                                                              
1e8ca836c962598551882e689265c1c5  Off with your head\nDance til you're dead\nHea...   
1296f487d135fa2808e18accebb6554f  Your body, your soul\nFeel the rhythm\nTake co...   
5a1fce1ebb19da07487f3132b261b78e  Last night, all I think about is you\nDon't st...   
a33e4b17ce2c355fb3e65ee042509803  Нас не догонят\nНас не догонят, нас не догонят...   
0e2b32a929c0e09569fdd0c831e2b459                                                NaN   
a6bea0f1418b5e768b689604e11d111d                                                NaN   
fb68073731d1fb23269c9778df343711                                                NaN   
96437a3517f1f408381363bc8d539718                                                NaN   
694ddd9c20051cd893222f00497b58a7     

Уже лучше. Для всех треков без метки есть какая-то информация, навание, текст, жанры и тд. Но для всех выведенных строк также пропущено и значения оригинального трека. А так как ранее мы выяснили, что все треки без оригинального `id` являются каверами, можно заполнить пропуски в этом столбце значением `COVER`.

In [50]:
df['track_remake_type'].fillna('COVER', inplace=True)

In [51]:
df.dropna(subset=['title'], inplace=True) #Удаляем строки с пропущенной информацией о треках

In [52]:
df.isna().sum() #проверяем

original_track_id    66504
track_remake_type        0
lyric_id             61098
text                 61098
dttm                     0
title                    0
language                 0
isrc                     0
genres                   0
duration                 0
dtype: int64

В одном из выводов замечаем, что столбец `genres` содержит пустые списки. Это не пропуски, поэтому они могли ускользнуть от внимания и впоследствии создать проблемы при обучении моделей. Посмотрим, сколько всего таких значений в датасете.

In [53]:
len(df[df['genres'].apply(lambda x: len(x) == 0)])

2315

Довольно много, удалять нельзя. Преобразуем списки в строки, и заменим их заглушкой `Unknown`

In [54]:
df = df.explode('genres')

In [55]:
df['genres'].replace('', 'Unknown', inplace=True)

In [56]:
df1 = df.query("isrc == 'PLS921752714'") #Проверяем, как заполнились пустые списки
df1

original_track_id track_remake_type lyric_id  \
track_id                                                                        
ffff8f1c01d78267ce067a2507fcc8f8              None             COVER      NaN   

                                 text                dttm     title language  \
track_id                                                                       
ffff8f1c01d78267ce067a2507fcc8f8  NaN 2019-05-17 16:50:33  No Limit       SQ   

                                          isrc genres  duration  
track_id                                                         
ffff8f1c01d78267ce067a2507fcc8f8  PLS921752714    NaN  175950.0

In [57]:
df['text'].fillna('', inplace=True)

Исследуем баланс классов целевого признака

In [65]:
df['track_remake_type'].value_counts()

COVER       99758
ORIGINAL     6649
Name: track_remake_type, dtype: int64

Видим дисбаланс 16 к одному, что закономерно, так как к одному треку может быть много каверов. Но для модели это может создать проблемы, при обучении нужно будет определить сбалансированные веса.

## Предобработка данных

Создадим новые фичи на основе векторизации текста.

In [58]:
# Параметры векторизации
max_features = 200
max_df = 0.80
min_df = 3

# Векторизация текста
vectorizer = TfidfVectorizer(lowercase=True, max_features=max_features, max_df=max_df, min_df=min_df)
matrix = vectorizer.fit_transform(df['text'])
sparse_matrix = csr_matrix(matrix)

# Создание позиций слов
Positions = {}
for j, word in enumerate(vectorizer.get_feature_names_out()):
    Positions[word] = j

# Количество слов
amountWords = len(vectorizer.get_feature_names_out())

# Создание DataFrame из TF-IDF матрицы
Tf_idf = pd.DataFrame.sparse.from_spmatrix(sparse_matrix, columns=vectorizer.get_feature_names_out())

# Применение PCA для уменьшения размерности
n_components = 30
pca = PCA(n_components=n_components)
matrix_pca = pca.fit_transform(sparse_matrix.toarray())

In [59]:
pca_df = pd.DataFrame(matrix_pca, columns=[f'PCA_{i}' for i in range(matrix_pca.shape[1])])
pca_df.index = df.index
df = pd.concat([df, pca_df], axis=1)

In [60]:
df

original_track_id  \
track_id                                                             
eeb69a3cb92300456b6a5f4162093851  eeb69a3cb92300456b6a5f4162093851   
eeb69a3cb92300456b6a5f4162093851  eeb69a3cb92300456b6a5f4162093851   
fe7ee8fc1959cc7214fa21c4840dff0a  fe7ee8fc1959cc7214fa21c4840dff0a   
fe7ee8fc1959cc7214fa21c4840dff0a  fe7ee8fc1959cc7214fa21c4840dff0a   
cd89fef7ffdd490db800357f47722b20  cd89fef7ffdd490db800357f47722b20   
...                                                            ...   
5e16b48a01483936ae404c3b30286a02                               NaN   
b8d4dc8f7abdd59df56eb39787b35c8f                               NaN   
b9233cb60674ca1ebb868e1cfe8d7b5c                               NaN   
b0e056b8cdd428972b758e8ea8a0ce8c                               NaN   
b0e056b8cdd428972b758e8ea8a0ce8c                               NaN   

                                 track_remake_type  \
track_id                                             
eeb69a3cb92300456b6a5f4162093851          ORIGINAL   
eeb69a3cb92300456b6a5f4162093851          ORIGINAL   
fe7ee8fc1959cc7214fa21c4840dff0a          ORIGINAL   
fe7ee8fc1959cc7214fa21c4840dff0a          ORIGINAL   
cd89fef7ffdd490db800357f47722b20          ORIGINAL   
...                                            ...   
5e16b48a01483936ae404c3b30286a02             COVER   
b8d4dc8f7abdd59df56eb39787b35c8f             COVER   
b9233cb60674ca1ebb868e1cfe8d7b5c             COVER   
b0e056b8cdd428972b758e8ea8a0ce8c             COVER   
b0e056b8cdd428972b758e8ea8a0ce8c             COVER   

                                                          lyric_id  \
track_id                                                             
eeb69a3cb92300456b6a5f4162093851  260f21d9f48e8de874a6e844159ddf28   
eeb69a3cb92300456b6a5f4162093851  260f21d9f48e8de874a6e844159ddf28   
fe7ee8fc1959cc7214fa21c4840dff0a  2498827bd11eca5846270487e4960080   
fe7ee8fc1959cc7214fa21c4840dff0a  2498827bd11eca5846270487e4960080   
cd89fef7ffdd490db800357f47722b20  5237001311d4062bf2b80de30652bf58   
...                                                            ...   
5e16b48a01483936ae404c3b30286a02                               NaN   
b8d4dc8f7abdd59df56eb39787b35c8f                               NaN   
b9233cb60674ca1ebb868e1cfe8d7b5c                               NaN   
b0e056b8cdd428972b758e8ea8a0ce8c                               NaN   
b0e056b8cdd428972b758e8ea8a0ce8c                               NaN   

                                                                               text  \
track_id                                                                              
eeb69a3cb92300456b6a5f4162093851  Left a good job in the city\nWorkin' for the m...   
eeb69a3cb92300456b6a5f4162093851  Left a good job in the city\nWorkin' for the m...   
fe7ee8fc1959cc7214fa21c4840dff0a  Some folks are born made to wave the flag\nOoh...   
fe7ee8fc1959cc7214fa21c4840dff0a  Some folks are born made to wave the flag\nOoh...   
cd89fef7ffdd490db800357f47722b20  Uno por pobre y feo, hombre\nPero antoja'o, ay...   
...                                                                             ...   
5e16b48a01483936ae404c3b30286a02                                                      
b8d4dc8f7abdd59df56eb39787b35c8f                                                      
b9233cb60674ca1ebb868e1cfe8d7b5c                                                      
b0e056b8cdd428972b758e8ea8a0ce8c                                                      
b0e056b8cdd428972b758e8ea8a0ce8c                                                      

                                                dttm            title  \
track_id                                                                
eeb69a3cb92300456b6a5f4162093851 2009-11-11 21:00:00       Proud Mary   
eeb69a3cb92300456b6a5f4162093851 2009-11-11 21:00:00       Proud Mary   
fe7ee8fc1959cc7214fa21c4840dff0a 2009-11-11 21:00:00    Fortunate Son   
fe7ee8fc1959cc7214fa21c4840

In [61]:
df.head(20) #Проверяем распаковку списков

original_track_id  \
track_id                                                             
eeb69a3cb92300456b6a5f4162093851  eeb69a3cb92300456b6a5f4162093851   
eeb69a3cb92300456b6a5f4162093851  eeb69a3cb92300456b6a5f4162093851   
fe7ee8fc1959cc7214fa21c4840dff0a  fe7ee8fc1959cc7214fa21c4840dff0a   
fe7ee8fc1959cc7214fa21c4840dff0a  fe7ee8fc1959cc7214fa21c4840dff0a   
cd89fef7ffdd490db800357f47722b20  cd89fef7ffdd490db800357f47722b20   
cd89fef7ffdd490db800357f47722b20  cd89fef7ffdd490db800357f47722b20   
cd89fef7ffdd490db800357f47722b20  cd89fef7ffdd490db800357f47722b20   
995665640dc319973d3173a74a03860c  995665640dc319973d3173a74a03860c   
995665640dc319973d3173a74a03860c  995665640dc319973d3173a74a03860c   
d6288499d0083cc34e60a077b7c4b3e1                              None   
4da9d7b6d119db4d2d564a2197798380                              None   
4da9d7b6d119db4d2d564a2197798380                              None   
2bf283c05b601f21364d052ca0ec798d                              None   
5eb13cb69b6e20dd7a42030f5936a9dc                              None   
5eb13cb69b6e20dd7a42030f5936a9dc                              None   
109f91266ef89cc3690079b28abfe9a3                              None   
109f91266ef89cc3690079b28abfe9a3                              None   
76c538125fc5c9ec6ad1d05650a57de5  76c538125fc5c9ec6ad1d05650a57de5   
76c538125fc5c9ec6ad1d05650a57de5  76c538125fc5c9ec6ad1d05650a57de5   
80a160ff31266be2f93012a2a3eca713  80a160ff31266be2f93012a2a3eca713   

                                 track_remake_type  \
track_id                                             
eeb69a3cb92300456b6a5f4162093851          ORIGINAL   
eeb69a3cb92300456b6a5f4162093851          ORIGINAL   
fe7ee8fc1959cc7214fa21c4840dff0a          ORIGINAL   
fe7ee8fc1959cc7214fa21c4840dff0a          ORIGINAL   
cd89fef7ffdd490db800357f47722b20          ORIGINAL   
cd89fef7ffdd490db800357f47722b20          ORIGINAL   
cd89fef7ffdd490db800357f47722b20          ORIGINAL   
995665640dc319973d3173a74a03860c          ORIGINAL   
995665640dc319973d3173a74a03860c          ORIGINAL   
d6288499d0083cc34e60a077b7c4b3e1             COVER   
4da9d7b6d119db4d2d564a2197798380             COVER   
4da9d7b6d119db4d2d564a2197798380             COVER   
2bf283c05b601f21364d052ca0ec798d             COVER   
5eb13cb69b6e20dd7a42030f5936a9dc             COVER   
5eb13cb69b6e20dd7a42030f5936a9dc             COVER   
109f91266ef89cc3690079b28abfe9a3             COVER   
109f91266ef89cc3690079b28abfe9a3             COVER   
76c538125fc5c9ec6ad1d05650a57de5          ORIGINAL   
76c538125fc5c9ec6ad1d05650a57de5          ORIGINAL   
80a160ff31266be2f93012a2a3eca713          ORIGINAL   

                                                          lyric_id  \
track_id                                                             
eeb69a3cb92300456b6a5f4162093851  260f21d9f48e8de874a6e844159ddf28   
eeb69a3cb92300456b6a5f4162093851  260f21d9f48e8de874a6e844159ddf28   
fe7ee8fc1959cc7214fa21c4840dff0a  2498827bd11eca5846270487e4960080   
fe7ee8fc1959cc7214fa21c4840dff0a  2498827bd11eca5846270487e4960080   
cd89fef7ffdd490db800357f47722b20  5237001311d4062bf2b80de30652bf58   
cd89fef7ffdd490db800357f47722b20  5237001311d4062bf2b80de30652bf58   
cd89fef7ffdd490db800357f47722b20  5237001311d4062bf2b80de30652bf58   
995665640dc319973d3173a74a03860c  e5b1b57090b728e8d98d2b4d9b781bf4   
995665640dc319973d3173a74a03860c  e5b1b57090b728e8d98d2b4d9b781bf4   
d6288499d0083cc34e60a077b7c4b3e1                               NaN   
4da9d7b6d119db4d2d564a2197798380  58b6145f2fb180f8cdc2067b4f1baebd   
4da9d7b6d119db4d2d564a2197798380  58b6145f2fb180f8cdc2067b4f1baebd   
2bf283c05b601f21364d052ca0ec798d  eb38211a25c1320991c5a23ad2417f33   
5eb13cb69b6e20dd7a42030f5936a9dc  486268c86660b734511cd7b67f70a2a8   
5eb13cb69b6e20dd7a42030f5936a9dc  486268c86660b734511cd7b67f70a2a8   
109f91266ef89cc3690079b28abfe9a3  a00a9ef94bcd5619e851f930dfc6ead7   
109f91266ef89cc3690079b28abfe9a3  a00a9ef94bcd5619e851f930d

In [62]:
df = df.reset_index()

In [64]:
df.isna().sum()

track_id                 0
original_track_id    99066
track_remake_type        0
lyric_id             90850
text                     0
dttm                     0
title                    0
language                 0
isrc                     0
genres                2315
duration                 0
PCA_0                    0
PCA_1                    0
PCA_2                    0
PCA_3                    0
PCA_4                    0
PCA_5                    0
PCA_6                    0
PCA_7                    0
PCA_8                    0
PCA_9                    0
PCA_10                   0
PCA_11                   0
PCA_12                   0
PCA_13                   0
PCA_14                   0
PCA_15                   0
PCA_16                   0
PCA_17                   0
PCA_18                   0
PCA_19                   0
PCA_20                   0
PCA_21                   0
PCA_22                   0
PCA_23                   0
PCA_24                   0
PCA_25                   0
P

Удаляем лишние признаки

In [66]:
df = df.drop(['original_track_id', 'lyric_id', 'isrc'], axis=1)

Создадим новый признак на основе даты (количество дней с момента релиза)

In [67]:
today = pd.to_datetime('2023-10-31')
df['days_since_release'] = (today - df['dttm']).dt.days

Создадим новый признак на основе названия (количество символов в названии)

In [68]:
df['title'] = df['title'].str.len()

Создадим новый признак на основе текста (количество слов в тексте)

In [69]:
df['text'] = df['text'].str.split().str.len()

In [70]:
df = df.drop(['dttm'], axis=1)

In [71]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 106407 entries, 0 to 106406
Data columns (total 38 columns):
 #   Column              Non-Null Count   Dtype  
---  ------              --------------   -----  
 0   track_id            106407 non-null  object 
 1   track_remake_type   106407 non-null  object 
 2   text                106407 non-null  int64  
 3   title               106407 non-null  int64  
 4   language            106407 non-null  object 
 5   genres              104092 non-null  object 
 6   duration            106407 non-null  float64
 7   PCA_0               106407 non-null  float64
 8   PCA_1               106407 non-null  float64
 9   PCA_2               106407 non-null  float64
 10  PCA_3               106407 non-null  float64
 11  PCA_4               106407 non-null  float64
 12  PCA_5               106407 non-null  float64
 13  PCA_6               106407 non-null  float64
 14  PCA_7               106407 non-null  float64
 15  PCA_8               106407 non-nul

In [72]:
df.head()

track_id track_remake_type  text  title language  \
0  eeb69a3cb92300456b6a5f4162093851          ORIGINAL   169     10       EN   
1  eeb69a3cb92300456b6a5f4162093851          ORIGINAL   169     10       EN   
2  fe7ee8fc1959cc7214fa21c4840dff0a          ORIGINAL   203     13       EN   
3  fe7ee8fc1959cc7214fa21c4840dff0a          ORIGINAL   203     13       EN   
4  cd89fef7ffdd490db800357f47722b20          ORIGINAL   348     15       ES   

    genres  duration     PCA_0     PCA_1     PCA_2  ...    PCA_21    PCA_22  \
0     ROCK  187220.0  0.467230  0.017720 -0.056349  ...  0.014357 -0.257105   
1  ALLROCK  187220.0  0.467230  0.017720 -0.056349  ...  0.014357 -0.257105   
2     ROCK  137780.0  0.284868 -0.007818  0.138148  ...  0.031384  0.012412   
3  ALLROCK  137780.0  0.284868 -0.007818  0.138148  ...  0.031384  0.012412   
4      POP  216840.0  0.022325 -0.070035  0.820291  ... -0.002297 -0.051246   

     PCA_23    PCA_24    PCA_25    PCA_26    PCA_27    PCA_28    PCA_29  \
0  0.125512  0.140414  0.005954  0.127955 -0.096559  0.175375 -0.075429   
1  0.125512  0.140414  0.005954  0.127955 -0.096559  0.175375 -0.075429   
2  0.057886  0.070817  0.001560  0.008068  0.051323  0.004052 -0.027370   
3  0.057886  0.070817  0.001560  0.008068  0.051323  0.004052 -0.027370   
4  0.004587 -0.019481  0.027739  0.012398 -0.045460 -0.010533  0.021776   

   days_since_release  
0                5101  
1                5101  
2                5101  
3                5101  
4                5152  

[5 rows x 38 columns]

Проведем разделение датафрейма на тестовую и обущающие выборки. Размер тестовой выборки возьмем 25%

In [74]:
df_train, df_test = train_test_split(df, test_size=0.25, random_state=RANDOM_STATE)

In [75]:
x_train = df_train.drop(['track_remake_type'], axis=1)
y_train = df_train[['track_remake_type']]
x_test = df_test.drop(['track_remake_type'], axis=1)
y_test = df_test[['track_remake_type']]

Заполним все пропуски заглушкой `Unknown`

In [76]:
x_train = x_train.fillna('Unknown')
x_test = x_test.fillna('Unknown')

Проведем кодирование категориальных признаков техникой `OrdinalEncoder`

In [77]:
encoder = OrdinalEncoder(handle_unknown='use_encoded_value', unknown_value=-1)

x_train_encoded = x_train.copy()
x_test_encoded = x_test.copy()

columns_to_encode = ['track_id' ,'language', 'genres']

x = pd.DataFrame(encoder.fit_transform(x_train_encoded[columns_to_encode]),
                               index=x_train_encoded.index, columns=columns_to_encode)

y = pd.DataFrame(encoder.transform(x_test_encoded[columns_to_encode]),
                              index=x_test_encoded.index, columns=columns_to_encode)

In [78]:
x_train_encoded.drop(columns_to_encode, axis=1, inplace=True)
x_test_encoded.drop(columns_to_encode, axis=1, inplace=True)

In [79]:
x_train_encoded = pd.merge(x_train_encoded, x, left_index=True, right_index=True)
x_test_encoded = pd.merge(x_test_encoded, y, left_index=True, right_index=True)

Целевой признак закодируем `LabelEncoder`

In [80]:
label_encoder = LabelEncoder()

y_train['track_remake_type'] = label_encoder.fit_transform(y_train['track_remake_type'])
y_test['track_remake_type'] = label_encoder.transform(y_test['track_remake_type'])

<ipython-input-80-0f19fb7710f4>:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  y_train['track_remake_type'] = label_encoder.fit_transform(y_train['track_remake_type'])
<ipython-input-80-0f19fb7710f4>:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  y_test['track_remake_type'] = label_encoder.transform(y_test['track_remake_type'])


Масштабируем численные признаки `StandartScaler`

In [81]:
numeric = ['text', 'title', 'duration', 'days_since_release']
scaler = StandardScaler()
scaler.fit(x_train[numeric])

x_train_scaled = pd.DataFrame(scaler.transform(x_train[numeric]))
x_test_scaled = pd.DataFrame(scaler.transform(x_test[numeric]))

x_train_scaled.index = x_train.index
x_test_scaled.index = x_test.index

In [82]:
x_train = pd.concat([x_train_scaled, x_train_encoded], axis=1)
x_test = pd.concat([x_test_scaled, x_test_encoded], axis=1)

In [83]:
x_train.columns = x_train.columns.astype(str)
x_test.columns = x_test.columns.astype(str)

In [84]:
y_train = np.ravel(y_train)
y_test = np.ravel(y_test)

In [85]:
x_train.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 79805 entries, 38605 to 15795
Data columns (total 41 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   0                   79805 non-null  float64
 1   1                   79805 non-null  float64
 2   2                   79805 non-null  float64
 3   3                   79805 non-null  float64
 4   text                79805 non-null  int64  
 5   title               79805 non-null  int64  
 6   duration            79805 non-null  float64
 7   PCA_0               79805 non-null  float64
 8   PCA_1               79805 non-null  float64
 9   PCA_2               79805 non-null  float64
 10  PCA_3               79805 non-null  float64
 11  PCA_4               79805 non-null  float64
 12  PCA_5               79805 non-null  float64
 13  PCA_6               79805 non-null  float64
 14  PCA_7               79805 non-null  float64
 15  PCA_8               79805 non-null  float64
 16  

In [86]:
x_train

0         1         2         3  text  title  duration  \
38605   1.296232 -0.207092 -0.853886 -0.204730   157     14  132590.0   
65483  -0.374019 -0.480441  0.404165 -0.455394     0     11  239200.0   
2732   -0.374019 -0.207092 -0.412429  2.931511     0     14  170000.0   
48457  -0.374019 -1.209370  0.378440 -0.316354     0      3  237020.0   
8703   -0.374019 -0.024860  1.104407  2.115873     0     16  298540.0   
...          ...       ...       ...       ...   ...    ...       ...   
54886  -0.374019 -0.207092 -0.312597 -0.466165     0     14  178460.0   
76820  -0.374019 -0.389325  0.626014 -0.523935     0     12  258000.0   
103694 -0.374019 -0.571557  0.068559 -0.839224     0     10  210760.0   
860     1.487726 -0.936022  0.204501  2.016978   175      6  222280.0   
15795  -0.374019 -0.571557  0.119891  0.790094     0     10  215110.0   

           PCA_0     PCA_1     PCA_2  ...    PCA_24    PCA_25    PCA_26  \
38605  -0.016615 -0.069573  0.749295  ...  0.024293  0.005693 -0.095462   
65483  -0.044260 -0.024796 -0.015149  ...  0.000304 -0.000173 -0.000330   
2732   -0.044260 -0.024796 -0.015149  ...  0.000304 -0.000173 -0.000330   
48457  -0.044260 -0.024796 -0.015149  ...  0.000304 -0.000173 -0.000330   
8703   -0.044260 -0.024796 -0.015149  ...  0.000304 -0.000173 -0.000330   
...          ...       ...       ...  ...       ...       ...       ...   
54886  -0.044260 -0.024796 -0.015149  ...  0.000304 -0.000173 -0.000330   
76820  -0.044260 -0.024796 -0.015149  ...  0.000304 -0.000173 -0.000330   
103694 -0.044260 -0.024796 -0.015149  ...  0.000304 -0.000173 -0.000330   
860     0.428772  0.015614 -0.001854  ... -0.070901  0.014882  0.012930   
15795  -0.044260 -0.024796 -0.015149  ...  0.000304 -0.000173 -0.000330   

          PCA_27    PCA_28    PCA_29  days_since_release  track_id  language  \
38605  -0.077106  0.038224 -0.056640                1058   16409.0      36.0   
65483  -0.000376  0.000297 -0.000535                 802   38289.0      16.0   
2732   -0.000376  0.000297 -0.000535                4261    5166.0      16.0   
48457  -0.000376  0.000297 -0.000535                 944   47116.0      17.0   
8703   -0.000376  0.000297 -0.000535                3428   52411.0      16.0   
...          ...       ...       ...                 ...       ...       ...   
54886  -0.000376  0.000297 -0.000535                 791   19212.0      17.0   
76820  -0.000376  0.000297 -0.000535                 732   58617.0      55.0   
103694 -0.000376  0.000297 -0.000535                 410   11161.0      68.0   
860    -0.022836  0.072506 -0.060545                3327   22676.0      59.0   
15795  -0.000376  0.000297 -0.000535                2074   53896.0      16.0   

        genres  
38605    107.0  
65483     96.0  
2732      96.0  
48457     68.0  
8703      52.0  
...        ...  
54886     68.0  
76820      1.0  
103694    29.0  
860       26.0  
15795     61.0  

[79805 rows x 41 columns]

## Обучение моделей

### Baseline

In [87]:
dummy = DummyClassifier(strategy='uniform')

dummy.fit(x_train, y_train)
y_dummy = dummy.predict(x_test)
dummy_f1 = f1_score(y_test, y_dummy)
print(f'F1 константной модели: {dummy_f1}')

F1 константной модели: 0.10976413229674528


Дамми-модель выдала метрику 0.1, все остальные модели должны быть выше неё, чтобы быть адекватными.

### Решающее дерево

In [88]:
%%time
model = DecisionTreeClassifier(random_state=RANDOM_STATE, class_weight='balanced')
params = { 'max_depth': [15] }
grid_tree = GridSearchCV(model, params, cv=5, n_jobs=-1,
                    scoring=make_scorer(f1_score))
grid_tree.fit(x_train, y_train)
print(grid_tree.best_params_)
tree_score = grid_tree.best_score_
tree_score

{'max_depth': 15}
CPU times: user 2.73 s, sys: 130 ms, total: 2.86 s
Wall time: 10.4 s


0.7024551672638457

У модели Решающего дерева метрика F1 существенно лучше

## Случайный лес

In [89]:
%%time
model = RandomForestClassifier(random_state=RANDOM_STATE, class_weight='balanced')
params = { 'n_estimators': [100] }
grid_forest = GridSearchCV(model, params, cv=5, n_jobs=-1,
                    scoring=make_scorer(f1_score))
grid_forest.fit(x_train, y_train)
print(grid_forest.best_params_)
random_forest_score = grid_forest.best_score_
random_forest_score

{'n_estimators': 100}
CPU times: user 23.8 s, sys: 196 ms, total: 24 s
Wall time: 1min 38s


0.7964408314842784

Модель случайного леса показывает существенно лучшую метрику F1

### Логистическая регрессия

In [90]:
%%time
model = LogisticRegression(random_state=RANDOM_STATE, class_weight='balanced')
params = { 'solver': ['liblinear'] }
grid_log = GridSearchCV(model, params, cv=5, n_jobs=-1,
                    scoring=make_scorer(f1_score))
grid_log.fit(x_train, y_train)
print(grid_log.best_params_)
logistic_regression_score = grid_log.best_score_
logistic_regression_score

{'solver': 'liblinear'}
CPU times: user 6.34 s, sys: 113 ms, total: 6.45 s
Wall time: 29.7 s


0.41474177642311433

Сильно хуже

### LGBM

In [96]:
%%time
model = LGBMClassifier(random_state=RANDOM_STATE, auto_class_weights='balanced')
params = { 'n_estimators': range(100, 500, 50),
           'max_depth': range(12, 24, 2),
           'learning_rate': [0.1]
         }
grid_lgb = GridSearchCV(model, params, cv=5, n_jobs=-1,
                    scoring=make_scorer(f1_score))
grid_lgb.fit(x_train, y_train)
print(grid_lgb.best_params_)
lgb_score = grid_lgb.best_score_
lgb_score = grid_lgb.best_score_

[LightGBM] [Warning] Unknown parameter: auto_class_weights
[LightGBM] [Warning] Accuracy may be bad since you didn't explicitly set num_leaves OR 2^max_depth > num_leaves. (num_leaves=31).
[LightGBM] [Warning] Unknown parameter: auto_class_weights
[LightGBM] [Warning] Accuracy may be bad since you didn't explicitly set num_leaves OR 2^max_depth > num_leaves. (num_leaves=31).
[LightGBM] [Info] Number of positive: 5007, number of negative: 74798
[LightGBM] [Info] Auto-choosing col-wise multi-threading, the overhead of testing was 0.028703 seconds.
You can set `force_col_wise=true` to remove the overhead.
[LightGBM] [Info] Total Bins 9829
[LightGBM] [Info] Number of data points in the train set: 79805, number of used features: 41
[LightGBM] [Info] [binary:BoostFromScore]: pavg=0.062740 -> initscore=-2.703954
[LightGBM] [Info] Start training from score -2.703954
{'learning_rate': 0.1, 'max_depth': 14, 'n_estimators': 450}
CPU times: user 16.3 s, sys: 1.59 s, total: 17.9 s
Wall time: 13min 

0.8291303141394604

Модель LGBM показывает лучший результат

## AdaBoost

In [99]:
%%time
model = AdaBoostClassifier(random_state=RANDOM_STATE)
params = { 'n_estimators': [100],
           'learning_rate': [0.1] }
grid_ada = GridSearchCV(model, params, cv=5, n_jobs=-1,
                    scoring=make_scorer(f1_score))
grid_ada.fit(x_train, y_train)
print(grid_ada.best_params_)
adaboost_score = grid_ada.best_score_
adaboost_score

{'learning_rate': 0.1, 'n_estimators': 100}
CPU times: user 22.4 s, sys: 246 ms, total: 22.6 s
Wall time: 1min 38s


0.6073213354486231

Результат плохой

### Stacking

### Тестирование

In [101]:
%%time
best_model = grid_lgb.best_estimator_
test_predictions = best_model.predict(x_test)
test_f1 = f1_score(y_test, test_predictions)
print(f'F1 на тестовой выборке: {test_f1}')

[LightGBM] [Warning] Unknown parameter: auto_class_weights
[LightGBM] [Warning] Accuracy may be bad since you didn't explicitly set num_leaves OR 2^max_depth > num_leaves. (num_leaves=31).
F1 на тестовой выборке: 0.8160882257541356
CPU times: user 745 ms, sys: 3.01 ms, total: 748 ms
Wall time: 750 ms


Получили результат 0.81

## Отчет по проекту

В проекте необходимо было провести выбор модели машинного обучения и подбор ее гиперпараметров с целью классификации музыкального трека на оригинальный и кавер.

**Основные этапы выполнения проекта:**

*1. Исследовательский анализ данных.*
- провели ознакомление с данными;
- изменили тип данных;
- определили категориальные признаки;
- проверили данные на пропуски;
- заполнили пропуски;
- выявили влияние категориальных признаков на целевой;
- выявили корреляцию между признаками;
- ислледовали признаки на мультиколлинеарность.

*2. Подготовка данных для моделей:*
- объединили все таблицы в одну;
- добавили признаки `days_since_release` (число дней с даты релиза трека по сегодняшнее время)
- провели векторизацию текстов песен с целью добавления новых признаков;
- ввели новый признак - количество слов в названии трека
- провели преобразование категориальных признаков в численные техниками OrdinalEncoding;
- провели преобразование целевого признака техникой LabelEncoding;
- стандартизировали числовые признаки техникой StandartScaler;
- разделили данные на 2 выборки обучающую и тестовую в пропорции 3:1;
- проверили дисбаланс признаков;

*3. Обучение моделей.*
- Baseline;
- LogisticRegression;
- RandomForestClassifier;
- LightGBM;
- выявили лучшую модель по метрике F1

Перед выполнением проекта был составлен план. По итогу выполнения проекта план был выполнен в полном объеме.


В процессе выполнения проекта встретились с:

- в столбце `original_track_id` есть почти 400 дубликатов. Видимо это как раз те оригиналы, на которых есть больше чем 1 кавер.
- выяснили, что все треки у которых отсутствует`original_track_id` являются каверами. На основании этого заполнили пропуски.
- повстречали довольно большое количество пропусков, большую часть из них получилось заполнить. Остальное удалили.
- в поле `dttm` были довольно странные значения. Разобрались, преобразовали их в дату выхода трека.
- на основании анализа названий треков провели детекцию языка исполнения и заполнили пропуски в столбце `language`

Для обучения моделей, оценки параметров и проведения оценки важности признаков были написаны функции, позволяющие работать с любой моделью и любыми входными данными. Это добавило универсальности коду.

В результате проведенной работы были обучены 5 моделей. На обучающих выборках методом кросс-валидации были определены значения F1:

|Модель|F1|
|:-----|:--------|
|LogisticRegression|0.40|
|RandomForestClassifier|0.8|
|LGBMClassifier|0.83|
|AdaBoost|0.61|

На основании тестировании модели на тестовых данных итоговое значение метрики F1 составило **0.81**

Т.е. модель с точностью в 81% предсказывает является музыкальный трек оригиналом или кавером. Такая точность позволит быть достаточно уверенным что запрос пользователя будет адекватно обработан.

Для решения оставшихся задач по группировке каверов и нахождения их оригиналов, а также создания цепочки каверов с целью выхода на оригинал выявили нехватку данных. Есть предположение, что если заполнить все пропуски в поле `text` и провести NLP обработку текстов, то на основе частотного анализа возможно с большой точностью с взять треки между собой, используя KNN (метод ближайших соседей).

Далее, создать алгоритм, который на основе проведенной группировки и с учетом признака кавер/оригинал и даты релиза создать цепочки треков.

Основной вопрос - где взять тексты песен. Мы не смогли найти рессурс откуда возможно спарсить эти данные и при этом не нарушить законодательство в области авторского права. Возможно, самый простой способ - заказчику применить технологию распознования речи и преобразования её в текст.